stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/controller/QueryChannelsController.kt

```diff
@@ -0,0 +1,391 @@
+package io.getstream.chat.android.offline.controller
+
+import io.getstream.chat.android.client.ChatClient
+import io.getstream.chat.android.client.api.models.FilterObject
+import io.getstream.chat.android.client.api.models.QuerySort
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.CidEvent
+import io.getstream.chat.android.client.events.MarkAllReadEvent
+import io.getstream.chat.android.client.events.NotificationAddedToChannelEvent
+import io.getstream.chat.android.client.events.NotificationMessageNewEvent
+import io.getstream.chat.android.client.events.UserPresenceChangedEvent
+import io.getstream.chat.android.client.events.UserStartWatchingEvent
+import io.getstream.chat.android.client.events.UserStopWatchingEvent
+import io.getstream.chat.android.client.logger.ChatLogger
+import io.getstream.chat.android.client.models.Channel
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.livedata.ChatDomainImpl
+import io.getstream.chat.android.livedata.controller.QueryChannelsController
+import io.getstream.chat.android.livedata.controller.QueryChannelsSpec
+import io.getstream.chat.android.livedata.extensions.users
+import io.getstream.chat.android.livedata.model.ChannelConfig
+import io.getstream.chat.android.livedata.request.QueryChannelsPaginationRequest
+import io.getstream.chat.android.livedata.request.toQueryChannelsRequest
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+
+private const val MESSAGE_LIMIT = 10
+private const val MEMBER_LIMIT = 30
+private const val INITIAL_CHANNEL_OFFSET = 0
+private const val CHANNEL_LIMIT = 30
+
+internal class QueryChannelsController(
+    val filter: FilterObject,
+    val sort: QuerySort<Channel>,
+    private val client: ChatClient,
+    private val domainImpl: ChatDomainImpl,
+) {
+    var newChannelEventFilter: (Channel, FilterObject) -> Boolean = { _, _ -> true }
+    var recoveryNeeded: Boolean = false
+
+    internal val queryChannelsSpec: QueryChannelsSpec = QueryChannelsSpec(filter, sort)
+
+    private val _channels = MutableStateFlow<Map<String, Channel>>(emptyMap())
+    internal val _loading = MutableStateFlow(false)
```

Comment:
I think all properties with "_" such prefix should be only private, and usually they're mutable. So if you want to expose loading flow then it's better to have something like:

``` kotlin
internal class QueryChannelsController {
 private val _loading = MutableStateFlow(false)
 val loading: Flow<Boolean> = _loading
}

```

Reply:
Because later we'll provide state flow versions to our customers and it's no good expose to them `_someValue`